### PR TITLE
Added back support for Hyperclick

### LINKED ISSUE #448: Implement hyperclick package
Part of Nuclide is the hyperclick package, https://atom.io/packages/hyperclick, it allows for interacting with text in the editor while holding the alt key, via a service.  

Some thoughts of things we could implement with this.
- Jump to definition.
- allow quick fixes to define syntax errors they are associated with
  - for example quick fix to implement an interface could be attached to the, the syntax error for a interface not being implemented.

### PR DESCRIPTION
**Pull request by sverrejoh**
Hyperclick support dissapeared after a refactor.

Added the old code back, and updated it to work with the new API, as
used in `go-to-definition`. Included support for selecting between
multiple definitions.

Should fix #448 and #1253

- [ ] All compiled assets are included (atom packages are git tags and hence the built files need to be a part of the source control)

### CODE REVIEW

**Comment by splodingsocks** [comment]
Awesome! I was really missing hyperclick. One issue I just found, though. If you use hyperclick to navigate to a symbol definition, `F10` doesn't return from the definition like it does when you use `F12` to go do the definition. It shows this error instead:

atom beta_2017-05-31 13-56-33

**Comment by sverrejoh** [comment]
Right, I saw some state related to that in the jump-to-definition code. This should probably be unified somehow, and I can see what I can do :). Thanks for the feedback!

**Comment by sverrejoh** [comment]
@splodingsocks Thank you again for the feedback! 

I fixed the issue with returning to definition with `F10` and updated the pull request. Should be more solid now.

**Comment by splodingsocks** [comment]
Nice!!

**Comment by sverrejoh** [comment]
@guncha Am I missing something in this pull request? I'm happy to clean it up if anything's missing.

**Comment by jonyeezs** [comment]
this should fix #1234 as well

File: lib/main/atom/hyperclickProvider.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import {simpleSelectionView} from "./views/simpleSelectionView"
+import {handleDefinitionResult} from "./commands/goToDeclaration"
+
+export function getHyperclickProvider(clientResolver: ClientResolver): any {
+    return {
+        providerName: "typescript-hyperclick-provider",
+        wordRegExp: /([A-Za-z0-9_])+|['"`](\\.|[^'"`\\\\])*['"`]/g,
+        getSuggestionForWord(editor: AtomCore.IEditor, text: string, range: TextBuffer.IRange) {
+            const TS_GRAMMARS = ["source.ts", "source.tsx"];
+            if (TS_GRAMMARS.indexOf(editor.getGrammar().scopeName) === -1) {
```
Comment (guncha) [inline lib/main/atom/hyperclickProvider.ts]: This should use `isTypescriptGrammar` helper from `typescriptEditorPane.ts`.

File: lib/main/atom/commands/goToDeclaration.ts [hunk 2]
```diff
@@ -56,3 +36,26 @@ commands.set("typescript:return-from-declaration", deps => {
       });
    }
 });
+
+export function handleDefinitionResult(result: protocol.DefinitionResponse,
+                                       location: FileLocationQuery): void {
+  if (result.body!.length > 1) {
```
Comment (sverrejoh) [inline lib/main/atom/commands/goToDeclaration.ts]: I agree. I just copied it, but then it was in a promise, so I guess we can argue the exception was handled :).

Fixed it now.

File: lib/main/atom/hyperclickProvider.ts (hunk 1 above)
Comment (sverrejoh) [inline lib/main/atom/hyperclickProvider.ts]: Great! I extracted it and put it in utils.

**Comment by guncha** [comment]
Solid PR! 👍 

I could have sworn I wrote a comment about using the scopes from the Typescript grammar to narrow down the locations where Hyperclick is effective. Basically, when you return from the function, the `range` property is used to highlight with an underline the section of code to let the user know "hey, if you click here, I'll take you somewhere". Currently, it will highlight every word, including string literals, and not everything, when clicked, will result in a jump to definition.

That's something that would take this from 95% to 100%, but it can be a part of a separate PR. I'll release this so people can start using it.

**Comment by sverrejoh** [comment]
Great suggestion 😊

I started looking into the scopes, and have something basic running. I'll try to come up with a list of what's not hyperclickable, and I can push it as a separate pull request. I think I need to add `scopeDescriptorForBufferPosition` to the TypeScript atom definition over at DefinitlyTyped for this as well.

From the top of my head:
- String literals outside import statement
- all other primitive types
- keywords
- within comments

Thanks a lot!

**Comment by guncha** [comment]
Sounds about right. You can always add it to the `typings/atom_core.d.ts` and then, if you want to be nice, open a PR against the DefinitelyTyped definitions.

**Pull request merged**